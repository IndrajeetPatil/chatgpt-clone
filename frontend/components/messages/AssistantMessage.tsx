import React, { useState } from "react";
import ReactMarkdown from "react-markdown";
import { Prism as SyntaxHighlighter } from "react-syntax-highlighter";
import {
  vs,
  vscDarkPlus,
} from "react-syntax-highlighter/dist/esm/styles/prism";
import FileCopyIcon from "@mui/icons-material/FileCopy";
import SmartToyIcon from "@mui/icons-material/SmartToy";
import {
  Box,
  IconButton,
  Paper,
  Tooltip,
  Typography,
  useTheme,
} from "@mui/material";

const CodeBlock: React.FC<{
  inline?: boolean;
  className?: string;
  children?: React.ReactNode;
}> = ({ inline, className, children }) => {
  const [codeCopied, setCodeCopied] = useState<string | null>(null);
  const theme = useTheme();

  const match = /language-(\w+)/.exec(className ?? "");
  const codeContent = String(children).replace(/\n$/, "");

  const handleCodeCopy = (code: string) => {
    navigator.clipboard.writeText(code);
    setCodeCopied(code);
    setTimeout(() => setCodeCopied(null), 2000);
  };

  if (!inline && match) {
    return (
      <Box position="relative" sx={{ mt: 2 }}>
        <SyntaxHighlighter
          style={theme.palette.mode === "dark" ? vscDarkPlus : vs}
          language={match[1]}
          PreTag="div"
        >
          {codeContent}
        </SyntaxHighlighter>
        <Tooltip title={codeCopied === codeContent ? "Copied!" : "Copy code"}>
          <IconButton
            onClick={() => handleCodeCopy(codeContent)}
            sx={{
              position: "absolute",
              top: 4,
              right: 4,
              backgroundColor:
                theme.palette.mode === "dark" ? "#4caf50" : "#ff9800",
              color: theme.palette.common.white,
              fontSize: 18,
              "&:hover": {
                backgroundColor:
                  theme.palette.mode === "dark" ? "#45a049" : "#e65100",
              },
            }}
          >
            <FileCopyIcon fontSize="small" />
          </IconButton>
        </Tooltip>
      </Box>
    );
  }

  return (
    <code
      style={{
        backgroundColor: theme.palette.mode === "dark" ? "#2d2d2d" : "#f5f5f5",
        borderRadius: "4px",
        padding: "0 4px",
        color: theme.palette.mode === "dark" ? "#e0e0e0" : "inherit",
      }}
    >
      {children}
    </code>
  );
};

interface AssistantMessageProps {
  content: string;
  isFirstMessage: boolean;
}

const AssistantMessage: React.FC<AssistantMessageProps> = ({
  content,
  isFirstMessage,
}) => {
  const [copied, setCopied] = useState(false);
  const theme = useTheme();

  const handleCopy = (text: string) => {
    navigator.clipboard.writeText(text);
    setCopied(true);
    setTimeout(() => setCopied(false), 2000);
  };

  const renderers = {
    code: CodeBlock,
  };

  return (
    <Box display="flex" justifyContent="flex-start" mb={2}>
      <Paper
        elevation={2}
        sx={{
          p: 2,
          backgroundColor:
            theme.palette.mode === "dark" ? "#2d2d2d" : "#fff3e0",
          maxWidth: "80%",
          wordWrap: "break-word",
          overflowWrap: "anywhere",
          position: "relative",
          display: "flex",
          flexDirection: "column",
        }}
      >
        <Box display="flex" alignItems="center" mb={1}>
          <SmartToyIcon
            sx={{
              color: theme.palette.mode === "dark" ? "#4caf50" : "#ff9800",
              fontSize: "1.5rem",
              mr: 1,
            }}
          />
          <Typography variant="body2" color="textSecondary">
            Assistant
          </Typography>
        </Box>
        <Typography variant="body1" component="div" sx={{ mt: 1 }}>
          <ReactMarkdown components={renderers}>{content}</ReactMarkdown>
        </Typography>
        {!isFirstMessage && (
          <Tooltip title={copied ? "Copied!" : "Copy entire message"}>
            <IconButton
              onClick={() => handleCopy(content)}
              sx={{
                position: "absolute",
                bottom: 4,
                right: 4,
                backgroundColor:
                  theme.palette.mode === "dark" ? "#4caf50" : "#ff9800",
                color: theme.palette.common.white,
                fontSize: 18,
                "&:hover": {
                  backgroundColor:
                    theme.palette.mode === "dark" ? "#45a049" : "#e65100",
                },
              }}
            >
              <FileCopyIcon fontSize="small" />
            </IconButton>
          </Tooltip>
        )}
      </Paper>
    </Box>
  );
};

export default AssistantMessage;
