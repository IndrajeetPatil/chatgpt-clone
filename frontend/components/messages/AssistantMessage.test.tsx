import React from "react";
import { render, screen } from "@testing-library/react";
import AssistantMessage from "./AssistantMessage";
import { ThemeProvider, createTheme } from "@mui/material";

// Mock modules
jest.mock("react-code-blocks", () => ({
  CopyBlock: ({ text }: { text: string }) => (
    <div data-testid="code-block">{text}</div>
  ),
  atomOneDark: {},
  atomOneLight: {},
}));

jest.mock("react-markdown", () => ({
  __esModule: true,
  default: ({
    children,
    components,
  }: {
    children: string;
    components: any;
  }) => {
    // Split content by code blocks and render appropriately
    const parts = children.split(/(```[\s\S]*?```)/);
    return (
      <div data-testid="markdown-content">
        {parts.map((part, index) => {
          if (part.startsWith("```")) {
            const Code = components.code;
            const codeContent = part.replace(/```\w*\n?|\n?```/g, "");
            return (
              <Code key={index} className="language-javascript">
                {codeContent}
              </Code>
            );
          }
          return <span key={index}>{part}</span>;
        })}
      </div>
    );
  },
}));

// Wrapper component for providing required context
const Wrapper = ({ children }: { children: React.ReactNode }) => (
  <ThemeProvider theme={createTheme()}>{children}</ThemeProvider>
);

describe("AssistantMessage", () => {
  test("renders plain text content", () => {
    const content = "This is a simple text message";
    render(<AssistantMessage content={content} isFirstMessage={false} />, {
      wrapper: Wrapper,
    });
    expect(screen.getByTestId("markdown-content")).toHaveTextContent(content);
  });

  test("renders code block content", () => {
    const content = '```javascript\nconsole.log("hello");\n```';
    render(<AssistantMessage content={content} isFirstMessage={false} />, {
      wrapper: Wrapper,
    });
    expect(screen.getByTestId("markdown-content")).toBeInTheDocument();
    expect(screen.getByTestId("code-block")).toBeInTheDocument();
  });

  test("renders inline code within text", () => {
    const content = "Use the `console.log()` function";
    render(<AssistantMessage content={content} isFirstMessage={false} />, {
      wrapper: Wrapper,
    });
    expect(screen.getByTestId("markdown-content")).toHaveTextContent(
      "Use the `console.log()` function"
    );
  });

  test("renders mixed content with code blocks and text", () => {
    const content =
      'Here is some code:\n```javascript\nconsole.log("hello");\n```\nAnd more text';
    const { container } = render(
      <AssistantMessage content={content} isFirstMessage={false} />,
      { wrapper: Wrapper }
    );

    expect(container).toMatchSnapshot();
  });

  test("handles empty content", () => {
    render(<AssistantMessage content="" isFirstMessage={false} />, {
      wrapper: Wrapper,
    });
    expect(screen.getByTestId("markdown-content")).toBeInTheDocument();
    expect(screen.getByTestId("markdown-content")).toHaveTextContent("");
  });
});
