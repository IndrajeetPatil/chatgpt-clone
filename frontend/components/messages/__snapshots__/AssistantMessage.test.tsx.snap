// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssistantMessage renders mixed content with code blocks and text 1`] = `
<div>
  <div
    class="MuiBox-root css-1w1vbxm"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 css-1bnegej-MuiPaper-root"
      style="--Paper-shadow: 0px 3px 1px -2px rgba(0,0,0,0.2),0px 2px 2px 0px rgba(0,0,0,0.14),0px 1px 5px 0px rgba(0,0,0,0.12);"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-vsuxn2-MuiSvgIcon-root"
        data-testid="SmartToyIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 9V7c0-1.1-.9-2-2-2h-3c0-1.66-1.34-3-3-3S9 3.34 9 5H6c-1.1 0-2 .9-2 2v2c-1.66 0-3 1.34-3 3s1.34 3 3 3v4c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2v-4c1.66 0 3-1.34 3-3s-1.34-3-3-3M7.5 11.5c0-.83.67-1.5 1.5-1.5s1.5.67 1.5 1.5S9.83 13 9 13s-1.5-.67-1.5-1.5M16 17H8v-2h8zm-1-4c-.83 0-1.5-.67-1.5-1.5S14.17 10 15 10s1.5.67 1.5 1.5S15.83 13 15 13"
        />
      </svg>
      <div
        class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
      >
        <div
          data-testid="markdown-content"
        >
          <span>
            Here is some code:

          </span>
          <div
            class="MuiBox-root css-1od93p9"
          >
            <div
              data-testid="code-block"
            >
              console.log("hello");
            </div>
          </div>
          <span>
            
And more text
          </span>
        </div>
      </div>
      <button
        aria-label="Copy entire message"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1tnv0fq-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element="true"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-120dh41-MuiSvgIcon-root"
          data-testid="ContentCopyIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12zm3 4H8c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h11c1.1 0 2-.9 2-2V7c0-1.1-.9-2-2-2m0 16H8V7h11z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
